### PR TITLE
COASTAL-460: Adds `.nextDate` and `.nextDateRepeating` to Alert facility

### PR DESCRIPTION
**Pull request by rickpasetto**
Adds ability to issue an alert that triggers on a specific date and time, given a `DateComponents`.  This effectively gives the facility the same functionality as UNCalendarNotificationTrigger (indeed, that's how it is implemented in Loop) (https://developer.apple.com/documentation/usernotifications/uncalendarnotificationtrigger).

https://tidepool.atlassian.net/browse/COASTAL-460

### CODE REVIEW

**Review by nhamming — APPROVED** [review]
LGTM

**Review by darinkrauss — APPROVED** [review]
Approving, but do we really want to support any known DateComponent? See TidepoolKit PR for details.

**Comment by ps2** [comment]
Like @darinkrauss , I'm also wanting to know if the use case requires this. Looking at the PR, I see this is an infusion assembly replacement reminder. For other reminders that happen on a fixed date, we have either a) scheduled for a fixed time in the future, or b) triggered the alert when the pump fires the alert, if this is a pump driven alert.

**Comment by rickpasetto** [comment]
@ps2 @darinkrauss I think that it is very reasonable for the Alert facility to support Alerts from Pumps, like the Infusion Assembly reminder, to be delivered at a specific time or date in a regular interval.  I don't see a reason to deviate from what the iOS Notification system has determined is a useful API.  Obviously, there isn't a logical reason for using some parts of `DateComponents` (e.g. `era`, `nanoseconds`), but I don't think there is a need to reinvent the wheel here.

**Comment by darinkrauss** [comment]
@rickpasetto How do you propose to limit the `DateComponents` to only the parts we support (particularly the parts the backend will support)?

**Comment by rickpasetto** [comment]
@darinkrauss technically iOS will support any `DateComponents`, and so will this Alert facility.  What does the backend need in terms of anything other than this object converted to JSON?  Maybe I'm not understanding the issue here...

**Comment by darinkrauss** [comment]
@rickpasetto The issue is that I don't think it appropriate for the backend to change its data model to support a whole bunch of less-than-useful options available on `DateComponents`. In particular, at first glance, I see a reasonable use case only for using (maybe) `month`, `day`, `hour`, `minute`, and (maybe) `second`.  The backend data model has always tried to stay agnostic, as much as possible, in terms of devices and OSes. In this case, I see no reason to include the other `DateComponent` options just to make it compatible with a specific OS data type.

So, how will the code or documentation ensure that some unsuspecting device manager developer doesn't accidentally use one of the unsupported `DateComponent` fields.

Also, we need to be explicit about the exact fields we support and provide a use case for each. (I'd rather not support `month` or `seconds` unless there is defined use case to do so, or, at the very least, a strongly suspected use case.)

Does that make sense?

**Comment by ps2** [comment]
Yes, I really do *not* want to invent more things to manage in Loop, even if they are just an abstraction on top of a platform facility.  We already have the ability to schedule at a particular date, right? By using the delayed trigger.  I would advise keeping the state space and amount of code as small as possible/necessary to meet our use cases.

**Comment by rickpasetto** [comment]
@darinkrauss and I chatted and I've agreed to make a specific type that only supports `days`, `hours`, and `minutes` for both `LoopKit` and `TidepoolKit`.  Stay tuned for the changes...

**Comment by rickpasetto** [comment]
@ps2 `delayed` does not really deal with time zone changes very well (say I want to be alerted 3 days from now at 8:00am, but I change time zones in that time...).  I think we want `DateComponents` functionality.

**Comment by ps2** [comment]
Assuming they do want to change the actual time of alert when they change timezones is not great. Generally we've kept the actual time that they want to be alerted. If they change timezones, the actual alert time shouldn't change without them knowing it.

**Comment by ps2** [comment]
Let's make sure the current setup absolutely doesn't work before inventing a new part of the loop system that has ripple effects over other models, and downstream maintenance effects. And yes, this is a new invention and an increase in state space/complexity for Loop even if it's using iOS system under the hood.

**Comment by rickpasetto** [comment]
@ps2 Right.  That's what I'm saying.  If the user chooses to be alerted in 3 days at 8:00am, and we calculate a `delayed` alert (which specifies a `TimeInterval` in number of seconds), we can of course calculate the number of seconds between `now` and 3 days from now at 8:00am and issue a delayed alert.  But say the user then gets on a plane and flies 3 time zones away.  That alert will not issue at 8am in their new time zone, it will issue at the wrong local time.  Does that make sense?  So, I do believe this is a required change.

**Comment by ps2** [comment]
No, that’s a different time. It might be the same wall clock time in a different time zone, but we are changing “when” the alert fires. If they want an alert 2 hours before the actual expiry, it will no longer be that.

**Comment by rickpasetto** [comment]
@ps2 that's not the requirement.  The requirement, as stated, and in agreement with how the Roche DM works, is a specific _time of day_ in a specific _number of days later_ (which I convert to a day of month).

**Comment by rickpasetto** [comment]
@ps2 that's not the requirement. The requirement, as stated, is a specific time of day in a specific number of days later (which I convert to a day of month).

I think this is a reasonable requirement. The hardware assembly has a certain lifespan... say 1-3 days, but isn't a hard requirement. The user then decides, "hey, remind me in 2 days at 8am, after I wake up, to replace it". This agrees with the need for a DateComponents - style approach.

**Comment by ps2** [comment]
Is there a hard limit that they eventually hit? Say they want to schedule a few hours before that? Then they travel -6 timezones. You're saying we're all of sudden, without their knowledge, adjust the time it fires by 6 hours just to make it match the wall clock in their new zone?  I agree that there might be assumptions the user makes about "8 o'clock is 8 o'clock, no matter what timezone I'm in, and that's what I said". But I think it's also reasonable to say that when they schedule the alert they're saying tues 8am *central*. And if they change timezones, it doesn't necessarily mean we change the actual time it fires, which we would be doing, if we try to follow wall clock. Also, in big travel adjustments, which way do you shift the time to match 8am? Forward or backward? Say you travel 11 hours of timezone change? Do you add 11 hours to the original time, or subtract 13?

**Comment by ps2** [comment]
I think the date component alert facility is great when the assumption of periodicity is heavily built in: I want alert every tues at 8am, for example. In this case, the date component periodicity assumption is pretty weak, as we're talking more about a time interval from a fairly random point in time: i.e. I want to change my set after 36 hours of use, a few hours before the actual expiration.  Just because you travel, you the actual lifetime usability of your component doesn't change, so I don't think we should assume that the user is super set on matching wall clock.

**Comment by rickpasetto** [comment]
Well, I will bring this up to product, but ultimately this will depend on the partner requirement, which will trump the decision.  I was hoping to avoid having to do that round trip delay, but if you want I can.

(Incidentally, in answering your "11 hour timezone change": in practice for this specific use case, we are always specifying a `day`, `hour`, and `minute`.  So, the alert would always match a future time.  I don't advocate making all 3 fields required, though, but if it will unblock this PR and the ticket, I'll do it...LMK).

**Comment by darinkrauss** [comment]
Thinking about this I see a potential problem with either `delayed` or `nextDate` that is not really resolved with either approach and product/partner needs to weight in on this. Let's say I want to be alerted X days later of an event, *but only at the time I specify* (though allowed/disallowed time range might be better).  For example, I want to be alerted at 8am two days from now informing me that the pump will expire at 11am (at least 3 hours before the alert, and I don't want to be woken up for something that is going to happen hours later after the alert is displayed). 

If we use 'delayed' with a hard-coded TimeInterval, then it will definitely alert at the technically "right" time (3 hours before pump expiration), but if I have changed timezones or DST, then it will potentially not be when I desired (latest 8am just prior to pump expiration). For example, if I travelled 6 timezones "west", then it will go off at 2am locally (when the pump expires at 8am locally). I suppose getting woken up is better than missing the alert, maybe? (Though having experienced alarm fatigue and interrupted sleep often enough, I'm not so sure about that.)

If we use `nextDate` with a hard-coded Date in the future (3 days from now at 8am), then it definitely won't alert when I don't want it to (e.g. 2am), but if I have changed timezones or DST, then it may be way before or *after* the actual event I'm worried about.  For example, if I travelled 6 timezones "east", then it will alert at 8am locally, *but the pump will already have expired at 5am locally*. As in this example, this gets complicated if the timezone change (6 hours) is more than the alert before the event (3 hours). And crossing the date line adds even more potential complication.

Ideally, the alert would reschedule itself within the specified allowed time range relative to the important event when any time or timezone or DSt change is detected.

Then again, perhaps I've entirely misunderstood the requirement here. :)

**Comment by rickpasetto** [comment]
I'm wondering if we're overthinking this.  Adding `nextDate` doesn't preclude clients from using `delayed`.  So, if they want that behavior (and want the concomitant complexity of dealing with time zone changes), they can choose to still use that.  

For the desired use case, the requirement is to provide the user the ability to choose a day, hour, and minute to be reminded of something.  I believe this achieves that, and assumes the user understands the consequences.

**Comment by darinkrauss** [comment]
I think that last part, "assumes the user understands the consequences", is what concerns me most. I'm not sure they will understand if their timezone change causes them to skip an important alert. At a minimum this seems like something that should be noted in the UI. In any case, I will defer to you and product/partner.

**Comment by rickpasetto** [comment]
I will bring this up internally with Product first, and see if the partner needs to be looped in.

**Comment by rickpasetto** [comment]
@ps2 @darinkrauss @nhamming I've discussed this with Product and the determination is that this requirement is analogous to a "meditation reminder", where the user really does want to be reminded at a specific time of day, even if they change time zones, and even if that time zone change misses the "first" time (alarm).  So, I think this is still going to be needed.

Do you have any other objections/comments other than that for this PR?

**Comment by ps2** [comment]
Yes, I do have objections. It's not analogous to a meditation reminder. It's a reminder about an action that needs to happen after a certain amount of time.

**Comment by rickpasetto** [comment]
After discussion with @ps2, product and partner, decided to not go forward with this change.

**Pull request closed without merging**
